File: src/agents/composer.py
"""
마리오네트 스튜디오 — Composer (작곡가) 에이전트
POST-PRODUCTION

global_audio_concept + 씬별 [Audio] 태그 기반으로 BGM 지시서 생성
Suno API 연동 시 실제 BGM 생성, 미연동 시 BGM 지시서(텍스트) 출력

현재: Gemini로 BGM 프롬프트 지시서 생성 (Suno API 연동 예정)
"""
import os
import json
import re
from dotenv import load_dotenv
from google import genai
from google.genai import types
from src.models.schemas import DirectionPlan


COMPOSER_SYSTEM_PROMPT = """You are a Film Score Composer at Marionette Studio.
Your job: Analyze the script's audio concept and each scene's [Audio] tag to design a detailed BGM/score brief.

For each scene, generate:
1. **mood**: emotional tone (e.g., "tension building", "melancholic", "triumphant")
2. **tempo**: BPM range (e.g., "90-100 BPM, slow burn")
3. **instruments**: key instruments (e.g., "cello ostinato, synth pad, glitchy percussion")
4. **reference**: real-world music reference (e.g., "Hans Zimmer - Time meets Trent Reznor")
5. **suno_prompt**: A ready-to-use prompt for Suno AI music generation (in English, 1-2 sentences)

RULES:
- Analyze the [Audio] tag in video_prompt for each scene
- Match the global_audio_concept tone
- Output in English
- suno_prompt should be concise and descriptive for AI music generation
"""


class ComposerAgent:
    """작곡가 — BGM/스코어 설계 및 생성"""

    def __init__(self, output_dir: str = "output/audio", api_key: str = None):
        load_dotenv(os.path.join(os.path.dirname(__file__), '..', '..', '.env'))
        self.api_key = api_key or os.getenv("Gemini_Api_Key") or os.getenv("GEMINI_API_KEY")
        self.output_dir = output_dir
        os.makedirs(self.output_dir, exist_ok=True)
        self.suno_api_key = os.getenv("SUNO_API_KEY", "")

        if self.api_key:
            self.client = genai.Client(api_key=self.api_key)
            self.model = "gemini-2.5-flash"
            print("🎵 Composer: Gemini 기반 BGM 설계 모드")
            if self.suno_api_key:
                print("   🎵 Suno API 감지 — 실제 BGM 생성 가능")
        else:
            self.client = None
            print("⚠️ Composer: API 키 미설정")

    def compose_score(self, json_path: str) -> str:
        """
        BGM 스코어 지시서 생성

        Returns: BGM 지시서 JSON 파일 경로
        """
        print("🎵 작곡가가 BGM 스코어를 설계합니다...")

        with open(json_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        plan = DirectionPlan(**data)

        # [Audio] 태그 추출
        scenes_info = ""
        for s in plan.scenes:
            audio_match = re.search(r'\[Audio\]\s*(.+?)$', s.video_prompt, re.IGNORECASE)
            audio_tag = audio_match.group(1).strip() if audio_match else "No specific audio direction"
            scenes_info += f"\nScene {s.scene_number} ({s.setting}, {s.time_of_day}): [Audio] {audio_tag}"

        user_prompt = f"""Design a film score brief for each scene.

Title: {plan.title}
Genre: {plan.genre}
Global Audio Concept: {plan.global_audio_concept}

Scenes:
{scenes_info}

Return a JSON array of objects with: scene_number, mood, tempo, instruments, reference, suno_prompt
"""

        try:
            response = self.client.models.generate_content(
                model=self.model,
                contents=user_prompt,
                config=types.GenerateContentConfig(
                    system_instruction=COMPOSER_SYSTEM_PROMPT,
                    response_mime_type="application/json",
                    response_schema={
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "scene_number": {"type": "integer"},
                                "mood": {"type": "string"},
                                "tempo": {"type": "string"},
                                "instruments": {"type": "string"},
                                "reference": {"type": "string"},
                                "suno_prompt": {"type": "string"},
                            },
                            "required": ["scene_number", "mood", "tempo", "instruments", "suno_prompt"],
                        },
                    },
                    temperature=0.6,
                ),
            )

            score_brief = json.loads(response.text)

            # 지시서 저장
            brief_path = os.path.join(self.output_dir, "score_brief.json")
            with open(brief_path, "w", encoding="utf-8") as f:
                json.dump(score_brief, f, ensure_ascii=False, indent=2)

            for item in score_brief:
                print(f"   🎵 S#{item['scene_number']}: {item['mood']} — {item['tempo']}")

            print(f"\n🎵 스코어 설계 완료! ({len(score_brief)}씬)")
            print(f"   📄 지시서: {brief_path}")
            return brief_path

        except Exception as e:
            print(f"❌ 작곡가 오류: {e}")
            return ""

File: apps/production-pipeline/packages/agents/src/post-production/composer.ts
// ---------------------------------------------------------------------------
// ComposerAgent — generates BGM/score for each scene using Suno API
// ---------------------------------------------------------------------------

import { mkdirSync, writeFileSync } from "node:fs"
import { join } from "node:path"
import type { DirectionPlan, Scene } from "@marionette/shared"
import { ProductionPhase } from "@marionette/shared"
import { BaseAgent } from "../base/agent.js"
import type { AgentInput, AgentOutput } from "../base/agent.js"

// ─── System prompt for music prompt generation ───

const COMPOSER_SYSTEM_PROMPT = `You are a world-class Film Composer at Marionette Studio.
Your job: Analyze each scene's mood, setting, and action, then create a precise music generation prompt.

For each scene (or scene group), generate a music prompt that includes:
1. **Genre/Style**: specific music genre (orchestral, electronic, ambient, jazz, etc.)
2. **Mood**: emotional tone (tense, melancholic, triumphant, eerie, etc.)
3. **Instruments**: key instruments (strings, synth pads, piano, percussion, etc.)
4. **Tempo**: BPM range (slow 60-80, moderate 80-120, fast 120-160)
5. **Dynamics**: volume/intensity arc (build-up, sustained, fade-out)

RULES:
- Output MUST be in English
- Consider the global_audio_concept for overall tone consistency
- Group consecutive scenes with similar moods into one BGM track
- Each prompt should be 2-3 sentences, vivid and descriptive
- Think Hans Zimmer meets Ryuichi Sakamoto meets Trent Reznor
- Output JSON array: [{ "group_label": "string", "scene_numbers": [1,2,3], "music_prompt": "string", "genre": "string", "duration_seconds": number }]`

// ─── Input extension ───

export interface ComposerInput extends AgentInput {
  directionPlan: DirectionPlan
  outputDir?: string
}

// ─── Music prompt entry ───

interface MusicPromptEntry {
  group_label: string
  scene_numbers: number[]
  music_prompt: string
  genre: string
  duration_seconds: number
}

// ─── Agent ───

export class ComposerAgent extends BaseAgent {
  readonly name = "Composer"
  readonly phase = ProductionPhase.POST
  readonly description =
    "Generates BGM/score tracks for scene groups using AI music generation (Suno)"

  async execute(input: ComposerInput): Promise<AgentOutput> {
    const { directionPlan, projectId, runId } = input
    const outputDir = input.outputDir ?? join("output", "audio", projectId)

    this.log(
      `Starting BGM composition for "${directionPlan.title}" (${directionPlan.scenes.length} scenes)`,
    )
    await this.updateProgress(runId, 5)

    mkdirSync(outputDir, { recursive: true })

    try {
      // Step 1: Generate music prompts via Gemini (scene analysis)
      const musicPrompts = await this.generateMusicPrompts(directionPlan)
      this.log(`Generated ${musicPrompts.length} music prompt groups`)
      await this.updateProgress(runId, 20)

      // Step 2: Generate BGM tracks via Suno for each group
      const generatedFiles: string[] = []

      for (let i = 0; i < musicPrompts.length; i++) {
        const group = musicPrompts[i]!
        this.log(
          `Generating BGM for "${group.group_label}" (scenes ${group.scene_numbers.join(",")})`,
        )

        const filePath = await this.generateBGM(
          group,
          outputDir,
          projectId,
          i,
        )

        if (filePath) {
          generatedFiles.push(filePath)

          // Save asset to DB for each scene in the group
          await this.saveAsset({
            projectId,
            type: "AUDIO",
            agentName: this.name,
            filePath,
            fileName: filePath.split("/").pop() ?? `bgm_group_${i}.mp3`,
            mimeType: "audio/mpeg",
            sceneNumber: group.scene_numbers[0],
            metadata: {
              groupLabel: group.group_label,
              sceneNumbers: group.scene_numbers,
              genre: group.genre,
              durationSeconds: group.duration_seconds,
              musicPrompt: group.music_prompt,
            },
          })
        }

        const progress = Math.round(20 + (70 * ((i + 1) / musicPrompts.length)))
        await this.updateProgress(runId, Math.min(progress, 95))
      }

      await this.updateProgress(runId, 100)

      this.log(
        `BGM composition complete (${generatedFiles.length} tracks from ${musicPrompts.length} groups)`,
      )

      return {
        success: true,
        message: `Composer generated ${generatedFiles.length} BGM tracks for "${directionPlan.title}"`,
        outputPath: outputDir,
        data: {
          generatedFiles,
          trackCount: generatedFiles.length,
          totalGroups: musicPrompts.length,
          musicPrompts,
        },
      }
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err)
      this.log(`Error during BGM composition: ${errorMessage}`)
      return {
        success: false,
        message: `Failed to compose BGM: ${errorMessage}`,
      }
    }
  }

  // ── Internal methods ──────────────────────────────────────────

  /** Use Gemini to analyze scenes and create music prompts grouped by mood. */
  private async generateMusicPrompts(
    plan: DirectionPlan,
  ): Promise<MusicPromptEntry[]> {
    const scenesText = plan.scenes
      .map(
        (s: Scene) =>
          `--- Scene ${s.scene_number} ---\nSetting: ${s.setting}\nTime: ${s.time_of_day}\nAction: ${s.action_description}\nDialogue: ${s.dialogue ?? "none"}`,
      )
      .join("\n\n")

    const userPrompt = `Analyze the scenes below and create music prompts grouped by similar mood/atmosphere.
Each group should cover 2-5 consecutive scenes. Aim for 3-8 total groups depending on scene count.

Title: ${plan.title}
Genre: ${plan.genre}
Global Audio Concept: ${plan.global_audio_concept}

${scenesText}`

    const response = await this.gateway.text(userPrompt, {
      provider: "gemini",
      systemPrompt: COMPOSER_SYSTEM_PROMPT,
      temperature: 0.6,
    })

    const cleaned = response.replace(/```json\n?|\n?```/g, "").trim()
    const parsed = JSON.parse(cleaned) as MusicPromptEntry[]

    return parsed
  }

  /** Generate a single BGM track via Suno API. */
  private async generateBGM(
    group: MusicPromptEntry,
    outputDir: string,
    projectId: string,
    groupIndex: number,
  ): Promise<string | null> {
    try {
      const audioBuffer = await this.gateway.audio(group.music_prompt, {
        provider: "suno",
        genre: group.genre,
        duration: group.duration_seconds,
        style: `Film score for ${group.group_label}`,
      })

      const filename = `bgm_${String(groupIndex + 1).padStart(2, "0")}_${group.group_label.replace(/\s+/g, "_").toLowerCase()}.mp3`
      const filePath = join(outputDir, filename)

      writeFileSync(filePath, audioBuffer)

      this.log(
        `BGM saved: ${filePath} (${Math.round(audioBuffer.length / 1024)}KB)`,
      )

      return filePath
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err)
      this.log(
        `BGM generation error for "${group.group_label}": ${errorMessage}`,
      )
      return null
    }
  }
}
